File: insertion_sort/javascript/insertion_sort.js
export default function sort(a) {
    if (!a) {
        return;
    }

    for (let i = 0; i < a.length; i += 1) {
        insert(a, i);
    }
}

function insert(a, i) {
    for (let j = i; j > 0 && a[j] < a[j - 1]; j -= 1) {
        swap(a, j, j - 1);
    }
}

function swap(a, i, j) {
    const t = a[i];
    a[i] = a[j];
    a[j] = t;
}

File: insertion_sort/go/insertion_sort.go
package insertion_sort

func Sort(a []int) {
	for i := 1; i < len(a); i++ {
		insertLast(a[:i+1])
	}
}

func insertLast(a []int) {
	for i := len(a) - 1; i > 0 && a[i] < a[i-1]; i-- {
		swap(a, i, i-1)
	}
}

func swap(a []int, i, j int) {
	t := a[i]
	a[i] = a[j]
	a[j] = t
}
